### ISSUE TITLE
'Keep last state' missing if using ngalerts

### ISSUE THREAD

**Issue by ivarkristian** [grafana/grafana#39056]
**What happened**:
After activating ngalerts and trying to setup an alert, in step 3 'Define alert conditions', the pulldown menu for what to do in case of execution error or timeout no longer contains the option 'Keep last state'. 

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.1.2 and 8.1.3
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

**Comment by dprokop** [comment]
@kylebrandt @peterholmberg could you folks have a look at this issue?

**Comment by peterholmberg** [comment]
some answers in here https://github.com/grafana/grafana/issues/36376

**Comment by ivarkristian** [comment]
I see that #36376 is closed. However, our use case for 'Keep last state' is that this allows us to separate between alerts triggered by the 'normal' alert conditions and alerts triggered by execution error. This is useful because alerts from normal alert conditions can be sent to people that are interested in the parameter that is measured, and alerts from execution error can be sent to people that knows their way around grafana/database/cloud system. If this is achievable in another way with ngalerts, please let me know! :)

**Comment by peterholmberg** [comment]
@gotjosh maybe a use case for Keep last state here?

**Comment by gotjosh** [comment]
Thanks @ivarkristian, we're certainly considering what to do with this area. But before going too much in detail - wouldn't

> Alert state if execution error or timeout

solve your problem?

**Comment by ivarkristian** [comment]
Great to know it is in the works, @gotjosh! Unfortunately, I fail to understand how using any of the current options in 'Alert state if execution error or timeout' would solve my problem. What I need, is the possibility to set up a 'normal' alert based on a regular alert condition, and send this to people that are interested in the parameter that triggers the alert. These people are not interested in execution errors, so I would like to set 'Alert state if execution error or timeout' to 'Keep last state'. With the current choices of 'Alerting' or 'OK', they will receive a notification if the execution error causes the alert state to change, which is confusing for many people.

For users that understand execution errors and are responsible for the system functionality, we configure a separate alert with 'Alert state if execution error or timeout' set to 'Alerting'. This works great as it is now, we just need 'Keep last state' to hide execution errors from the non-techies.

Instead of reintroducing Keep last state, another way that could be more in line with the new alerting system, is to have alert reason (query/execution/nodata etc) as a meta data in the alert. In that way we could filter alerts based on alert reason when we configure notification policies.

**Comment by mckn** [comment]
@gotjosh are we waiting for more information on this issue or should that label be removed? We are doing follow up on all issues with the "needs more info" on a weekly basis and it seems like that information is provided in this issue.

**Comment by gerobinson** [comment]
I think this use case will be supported once https://github.com/grafana/grafana/pull/40540 and https://github.com/grafana/grafana/pull/41869 are released in 8.3?

**Comment by gerobinson** [comment]
Hi @ivarkristian,

I think this use case will be supported in Grafana 8.3 via the `DatasourceNoData` and `DatasourceError` alerts that replace keep last state.

**Comment by gerobinson** [comment]
Hi @ivarkristian,

I think this use case will be supported in Grafana 8.3 via the DatasourceNoData and DatasourceError alerts that replace keep last state.

**Comment by gotjosh** [comment]
Both of these have merged and are included in the latest beta of 8.3 (8.3-beta2 at the moment of writing).

Let us know your thought on whether the new approach solves your problem. The details of what the what options are available and what they do are in: https://grafana.com/docs/grafana/next/alerting/unified-alerting/alerting-rules/create-grafana-managed-rule/#no-data-and-error-handling

**Comment by danielfaust** [comment]
> Both of these have merged and are included in the latest beta of 8.3 (8.3-beta2 at the moment of writing).

Yesterday I upgraded from 8.3.3 to 9.2.5 and now all my alerts are firing like crazy and am in this exact same problem that the "Keep last state"-functionality has disappeared.

For example: I have a data source which sends data to InfluxDB at a random interval, from 30 minutes down to 2 seconds. The data is a value between 0.0 and 1.0.

The alert should fire when it is below 0.5 and resolve if equal or above.

Every 10 seconds it samples the last 10 seconds and creates an average of the found values and that average is used to determine if it should fire or not.

Upon state change, it triggers a webhook which checks the "state" of "alerting" or "ok" in order to send an adequate message via email.

This used to work perfectly with 8.3.3, but now, whenever it enters into a NoData condition and then a new data packet arrives, it will toggle the state from "alerting" to "ok", disregarding the actual condition of the value, and the next time this happens then toggle from "ok" to "alerting", again disregarding the value.

So if the value is at 0.1, clearly in the "alerting" region, if a NoData condition disappears because a new value comes in (0.1), it will toggle from the previous "ok" or "alerting" to the opposite value, regardless of it needing to be always "alerting", because the value is of 0.1. Also, since that new data packet is still in the "alerting" range. there is actually no reason to re-trigger the alert.

So now I'm constantly getting notifications, half of them erroneously telling me the wrong state, just because a new data packet came in. Not only is this an issue, but now also the graph is full of vertical lines showing when alerts have triggered (unneccesarily)

KeepLastState should be an option and as long as the state has not changed, no notification should be sent. And NoData should *not* be treated as a state, but as another notification type which could help to discover outages in the data source, but which does not affect the actual state of the alarm.

Also, the Grafana UI enters an erroneous state because the `endsAt` is lower than `startsAt`, when the alarm enters an "alerting" state while the Alert tab is open.

In the following case it's a Test alert, but the actual alerts look the same when firing:

```
{
  "receiver": "",
  "status": "firing",
  "alerts": [
    {
      "status": "firing",
      "labels": {
        "alertname": "TestAlert",
        "instance": "Grafana"
      },
      "annotations": {
        "summary": "Notification test"
      },
      "startsAt": "2022-11-21T09:07:02.116297869Z",
      "endsAt": "0001-01-01T00:00:00Z",                        <<<<<<<<<<<<<<<<<<<<<
      "generatorURL": "",
      "fingerprint": "57c6d9296de2ad39",
      "silenceURL": "http://localhost:3000/alerting/silence/new?alertmanager=grafana&matcher=alertname%3DTestAlert&matcher=instance%3DGrafana",
      "dashboardURL": "",
      "panelURL": "",
      "valueString": "[ metric='foo' labels={instance=bar} value=10 ]"
    }
  ],
  "groupLabels": {},
  "commonLabels": {
    "alertname": "TestAlert",
    "instance": "Grafana"
  },
  "commonAnnotations": {
    "summary": "Notification test"
  },
  "externalURL": "http://localhost:3000/",
  "version": "1",
  "groupKey": "{alertname=\"TestAlert\", instance=\"Grafana\"}2022-11-21 09:07:02.116297869 +0000 UTC m=+55809.599736604",
  "truncatedAlerts": 0,
  "orgId": 1,
  "title": "[FIRING:1]  (TestAlert Grafana)",
  "state": "alerting",
  "message": "**Firing**\n\nValue: [ metric='foo' labels={instance=bar} value=10 ]\nLabels:\n - alertname = TestAlert\n - instance = Grafana\nAnnotations:\n - summary = Notification test\nSilence: http://localhost:3000/alerting/silence/new?alertmanager=grafana&matcher=alertname%3DTestAlert&matcher=instance%3DGrafana\n"
}
```


```
An unexpected error happened
Details
RangeError: The start of an interval cannot be after its end

    at g (http://10.10.20.24:3000/public/build/7062.5ec1c4d918175e7b8ce4.js:68:179)
    at div
    at div
    at div
    at u (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:7036:171)
    at div
    at I (http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:45:57)
    at div
    at div
    at div
    at t (http://10.10.20.24:3000/public/build/4713.232e2ff90daf715727af.js:2:681363)
    at p (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:420:3279)
    at Ro (http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:363:55)
    at div
    at d (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:2935:30)
    at div
    at http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:426:1562
    at div
    at div
    at n (http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:99:6538)
    at div
    at n (http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:99:8487)
    at div
    at n (http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:99:6538)
    at div
    at n (http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:99:8487)
    at Zt (http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:99:14074)
    at div
    at div
    at div
    at b (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:5803:1142)
    at Si (http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:443:1519)
    at g (http://10.10.20.24:3000/public/build/4713.232e2ff90daf715727af.js:2:999564)
    at WithTheme(Connect(Si))
    at Na (http://10.10.20.24:3000/public/build/DashboardPage.1c474db1e4606824e351.js:566:7310)
    at DashboardPage
    at g (http://10.10.20.24:3000/public/build/4713.232e2ff90daf715727af.js:2:999564)
    at Suspense
    at c (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:1194:301)
    at kn (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:5514:360)
    at t (http://10.10.20.24:3000/public/build/4713.232e2ff90daf715727af.js:2:1014882)
    at t (http://10.10.20.24:3000/public/build/4713.232e2ff90daf715727af.js:2:1016903)
    at main
    at st (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:5475:13054)
    at t (http://10.10.20.24:3000/public/build/4713.232e2ff90daf715727af.js:2:1011500)
    at div
    at o (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:2035:4592)
    at t.KBarProvider (http://10.10.20.24:3000/public/build/4713.232e2ff90daf715727af.js:2:261503)
    at l (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:6236:19180)
    at c (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:1194:301)
    at u (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:1194:855)
    at y (http://10.10.20.24:3000/public/build/4713.232e2ff90daf715727af.js:2:2892432)
    at u (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:6229:10872)
    at l (http://10.10.20.24:3000/public/build/4713.232e2ff90daf715727af.js:2:997609)
    at ss (http://10.10.20.24:3000/public/build/3191.553394791727cd6a7d14.js:5527:167)
```
